### DISCUSSION TITLE
MASSIVE Ram consumption

### DISCUSSION THREAD

**Discussion by Edgarborras94** [livewire/livewire#3086] [7. Bugs (archived)]
### Description

Was creating a CRUD system without the use of a controller, using exclusivelly livewire. I put 3 fields and when click a button, it fires a method in the livewire Component to create that model, wiping those fields and without refreshing the page.

All runs perfectly, but, when I insert tens of them it starts to consume ALOT of ram (2-4GB).

I added pagination to the component to reduce the size of the page, but it didn't help.

### Exact steps to reproduce

I don't know how to reproduce, but I will share part of my code, maybe I did something wrong? (rushing is not good)

### Stripped-down, copy-pastable code snippets

####blade:
Modal to create the models:
```php

<div @xshow('showModalCreateProducto', 'tailwindui.modal.panel')>                
    <div>
        <div>
            <h3  id="modal-headline">Añadir Producto</h3>
            <div class="mt-2">
                <div>
                    <label>Categoría del Producto</label>
                    <input wire:model="nuevaCategoria">
                </div>
                <div>
                    <label>Nombre del producto</label>
                    <input wire:model="nuevoNombre" >
                </div>
                <div>
                    <label>Precio del producto</label>
                    <input wire:model="nuevoPrecio" >
                </div>
            </div>
        </div>
    </div>
    <div>
        <button wire:click="createProducto()" type="submit">Añadir</button>
        <button @click="showModalCreateProducto = false" type="button">Cancelar</button>
    </div>
</div>

```
Method "createProducto" from component:
```php
class CartaAdmin extends Component
{
    use WithPagination;

    public Empresa $empresa;

    public $nombres;
    public $precios;
    public $categorias;

    public $nuevoNombre;
    public $nuevoPrecio;
    public $nuevaCategoria;

    public function createProducto(){
        $validatedData = $this->validate([
            'nuevoNombre' => 'required',
            'nuevoPrecio' => 'required|numeric',
        ]);

        $producto = new Producto();
        $producto->nombre = $this->nuevoNombre;
        $producto->precio = $this->nuevoPrecio;
        $producto->empresa_id = $this->empresa->id;
        if($this->nuevaCategoria !== null){
            $categoria = CategoriaProducto::where([['name',$this->nuevaCategoria],['empresa_id',$this->empresa->id]])->first();
            if($categoria !== null){
                $categoriaSeleccionada = $categoria;
            }
            else{
                $categoriaSeleccionada = new CategoriaProducto();
                $categoriaSeleccionada->name = $this->nuevaCategoria;
                $categoriaSeleccionada->empresa_id = $this->empresa->id;
                $categoriaSeleccionada->save();
            }
            $producto->categoria_producto_id = $categoriaSeleccionada->id;
        }
        $producto->save();
        unset($this->nuevoNombre);
        unset($this->nuevoPrecio);
        unset($this->nuevaCategoria);
        $this->resetPage();
    }
/* More ommited code */
}
```

And then I paginate the products to the view:
```php
public function render()
    {
        return view('livewire.carta-admin', [
            'productos' => $this->empresa->productos()->orderBy('categoria_producto_id')->paginate(15),
        ]);
    }
```

And then those products are displayed in a complex table with nested rows and edit buttons for each product

**Context**
- Livewire version: 2.3.6
- Laravel version: 8.2.2
- Alpine version: 2.7.3
- Browser: Chrome 87

**Comment by marispro** [comment]
Please share your project to be available to test. It is impossible to test it without database data.

I usually check how much data consumes AJAX requests. If it is more than 30 KB, then something is wrong and should be optimised (by wire:ignore or something else). 

PS: I am sorry,  but need to clarify - is it browser consuming much data or your web server?

**Comment by stevebauman** [comment]
In addition to the above comment, @Edgarborras94 could you also please re-format your code samples with readable spacing? It would really help reduce the mental load and increase readability, making it easier for us to help you out ❤️

**Comment by Edgarborras94** [comment]
Hi @marispro , the ram consumption is from my browser. Here you have a test user to check:
url: https://instantpay.es/admin/cartas/3
user: TestLivewire
password: gpbb@xEDG2U3

The problem can be replicated with the "Añadir nuevo Producto" button.

I will edit my code @stevebauman sorry about that 😋

**Comment by marispro** [comment]
Hi @Edgarborras94, I can't log in, it says: Estas credenciales no coinciden con nuestros registros.

**Comment by Edgarborras94** [comment]
@marispro Sorry about that, fixed, try now and thanks in advance!
![image](https://user-images.githubusercontent.com/6952403/105399212-9f3b3380-5c23-11eb-83bd-2b77095e93c1.png)
This is a screenshot I took the other day 😅

**Comment by marispro** [comment]
I tried but browser ram still looks fine. Can you check that exactly this page makes your RAM overload?
Check chrome task manager and then sort by CPU
![image](https://user-images.githubusercontent.com/5610414/105400221-eece2f00-5c24-11eb-8a26-f62261e97412.png)

**Comment by Edgarborras94** [comment]
thanks for that tip @marispro ! I noticed that if I keep the tab opened in the front layer, ram consumption increases by time, I use a wire:poll for the table... maybe that has something to do with that? 🤔🤔
![image](https://user-images.githubusercontent.com/6952403/105402621-edeacc80-5c27-11eb-82ba-d143e09fe84b.png)

I will keep that tab open while I have dinner. Again, thanks a lot!

**Comment by Edgarborras94** [comment]
This is what I got after the dinner break: (without using the page)
![image](https://user-images.githubusercontent.com/6952403/105414861-59d53100-5c38-11eb-8e09-317b3ffd822b.png)
And this is what I get after several insertions in a row
![image](https://user-images.githubusercontent.com/6952403/105415191-bd5f5e80-5c38-11eb-971c-414226a042a7.png)
1.5GB LOL

maybe the result of the other day were because I inserted like 50 items in a row. I don't get where is the problem, should I update livewire or something? maybe changing the wire:poll aproach?

EDIT: OMG
![image](https://user-images.githubusercontent.com/6952403/105416321-3dd28f00-5c3a-11eb-9652-ee4b28a3f59a.png)
im gonna try to reduce the html of the table, try another browser maybe... this thing is getting me crazy lol

**Comment by joshhanley** [comment]
@Edgarborras94 can you share your table html? Maybe it's a morphdom issue. Do you have any nested components in the table? Also make sure you have keys set correctly on your table rows in your foreach loop, and on any child components if you have them.

I suspect either new components are getting added but the old ones aren't getting removed or something similar with the transmitted html.

**Comment by Edgarborras94** [comment]
Sure @joshhanley here you have:

```php
<table x-data="{@foreach($empresa->categoriaProductos as $categoria) 'showDesplegable{{$categoria->id}}': true, @endforeach}">
    <thead>
        <tr>
            <th scope="col">Nombre</th>
            <th scope="col">Precio</th>
            <th scope="col"></th>
        </tr>
    </thead>
    @foreach($empresa->categoriaProductos as $categoria)
        @if($productos->where('categoria_producto_id',$categoria->id)->count() > 0)
            <tbody>
                <tr>
                    <td colspan="2">
                            <button @click="showDesplegable{{$categoria->id}}=!showDesplegable{{$categoria->id}}" > </button>
                            <strong>{{$categoria->name}}</strong>
                    </td>
                    <td>
                    @if($categoria->productos->count() == 0)
                        <button wire:click="deleteCategoria({{$categoria->id}})" type="button" >
                            <i class="fa-thin fa-trash-can mr-2"></i>
                            Eliminar Categoría
                        </button>
                    @endif
                    </td>
                </tr>
            </tbody>
        @endif
        <tbody x-show="showDesplegable{{$categoria->id}}">
            @foreach($productos->where('categoria_producto_id',$categoria->id) as $producto)
                <tr>
                    <td>{{$producto->nombre}}</td>
                    <td>{{number_format($producto->precio, 2)}} €</td>
                    <td wire:ignore>
                        <button wire:click="editarProducto({{$producto->id}})" type="button" >
                            Editar
                        </button>
                    </td>
                </tr>
            @endforeach
        </tbody>
    @endforeach
</table>
```
